apps/designer/app/designer/features/style-panel/controls/css-property/css-property.tsx

```diff
@@ -0,0 +1,246 @@
+import {
+  Button,
+  Combobox,
+  DropdownMenu,
+  DropdownMenuContent,
+  DropdownMenuPortal,
+  DropdownMenuRadioGroup,
+  DropdownMenuRadioItem,
+  DropdownMenuTrigger,
+  Flex,
+  IconButton,
+  numericScrubControl,
+  TextField,
+} from "@webstudio-is/design-system";
+import { ChevronDownIcon } from "@webstudio-is/icons";
+import {
+  KeywordValue,
+  StyleProperty,
+  StyleValue,
+  Unit,
+  units,
+} from "@webstudio-is/react-sdk";
+import { useCombobox } from "downshift";
+import { KeyboardEvent, PointerEvent, useCallback, useRef } from "react";
+import { PropertyIcon } from "../../shared/property-name";
+
+const sortedUnits = units
+  .slice(0)
+  .sort((v) =>
+    ["%", "px", "rem", "em", "ch", "vh", "vw", "hv", "vmin", "vmax"].includes(v)
+      ? -1
+      : 1
+  );
+
+type CssPropertyControlProps = {
+  property: StyleProperty;
+  value?: StyleValue;
+  allowedValues?: Array<KeywordValue>;
+  onChange?: (value?: StyleValue) => void;
+  onChangeComplete?: (value?: StyleValue) => void;
+};
+
+export const CssPropertyControl = ({
+  property,
+  value,
+  allowedValues = [],
+  onChange,
+  onChangeComplete,
+}: CssPropertyControlProps) => {
+  const inputRef = useRef<HTMLInputElement | null>(null);
+
+  const handleArrowUpDown = (event: KeyboardEvent<HTMLInputElement>) => {
+    // @todo: There is a bug when type === "keyword" and the user presses arrow up/dow then it's going to be displayed NaN as a value
+    if (value?.type === "unit") {
+      let numberValue = parseFloat(event.currentTarget.value);
+      const direction = event.code === "ArrowUp" ? 1 : -1;
+      let currentDelta = 1;
+      if (event.shiftKey) currentDelta = 10;
+      if (event.altKey) currentDelta = 0.1;
+      numberValue = numberValue + currentDelta * direction;
+      const roundedValue =
+        numberValue % 1
+          ? Number(
+              numberValue.toPrecision(
+                Math.abs(numberValue).toString().indexOf(".") + 2
+              )
+            )
+          : numberValue;
+      event.currentTarget.value = String(roundedValue);
+      onChange?.({
+        type: "unit",
+        value: roundedValue,
+        unit: value?.unit ?? "px",
+      });
+    }
+  };
+
+  const unitStateReducer = useCallback((state, actionAndChanges) => {
```

Comment:
can't this be a separate function outside of component? I personally really really don't like huge components